api/python/provisioner/commands/validator/validator.py

```diff
@@ -0,0 +1,250 @@
+import logging
+from abc import abstractmethod
+from pathlib import Path
+from typing import Dict, Callable
+
+from ...errors import ValidationError
+
+logger = logging.getLogger(__name__)
+
+
+class SchemeValidator:
+
+    """Base Scheme Validator. It defines the interface for child classes."""
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def validate(self, *args, **kwargs):
+        """
+        Abstract Scheme validation method.
+
+        Parameters
+        ----------
+        args
+            list of positional arguments
+        kwargs
+            keyword-only arguments
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        ValidationError
+            If validation is failed.
+        """
+        pass
+
+
+class Validator:
+
+    """Abstract Validator class defines the interface for inheritance."""
+
+    @abstractmethod
+    def validate(self, *args, **kwargs):
+        """
+        Abstract validation method of Validator.
+
+        Parameters
+        ----------
+        args
+            list of positional arguments
+        kwargs
+            keyword-only arguments
+
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        ValidationError
+            If validation is failed.
+        """
+        pass
+
+
+class FileValidator(Validator):
+
+    """Class for file validation."""
+
+    def __init__(self, *, required: bool = False,
+                 content_validator: Callable[[Path], None] = None):
+        """
+        File validator initialization method
+
+        Parameters
+        ----------
+        required: bool
+            if True validation raises an Exception if the file doesn't exist
+        content_validator: Callable[[Path], None]
+            callable object for file content validation
+
+        Notes
+        -----
+        TBD: other possible parameters
+        permissions: int/str
+            requested file permissions
+        owner: str
+            requested file owner
+        group: str
+            requested file group
+        is_symlink: bool
+            defines if file should be a symlink
+        """
+        self._required = required
```

Comment:
I would suggest to use `attr` helpers here like in other API parts:
- please use `attr.ib` (if actually needed) to declare attrs
- for all validator you may actually use attr validators so validation would during both validator object creation time and during `validator` method call
- many of validations you do here are already available: attr [built-in validators](https://www.attrs.org/en/stable/api.html#validators) and provisioner custom [ones](https://github.com/Seagate/cortx-prvsnr/blob/bf833997fa2bd383e9109f5e748538c04acab327/api/python/provisioner/utils.py#L44-L72))
- for more complicated validators you can write your own as a separate function in the module
- also you can consider to use [converters](https://www.attrs.org/en/stable/init.html#converters) and make you input assumptions more flexible (e.g. `Union[str, Path]`), like [here](https://github.com/Seagate/cortx-prvsnr/blob/bf833997fa2bd383e9109f5e748538c04acab327/api/python/provisioner/paths.py#L27)
- docstring can be attached to class level summary, like [here](https://github.com/rackerlabs/otter/blob/ec0a76d25dcbb50636dfc9316674715e138b9d0c/otter/convergence/selfheal.py#L27-L58)

To summarize:
- for object level attributes (`__init__` args) I suggest to use `attr` API
- for `validator` method you can re-use the same validator functions (isinstance, Path exists ...)
- converters would help you as well

Finally the module (currently about 250 lines) might be condensed into 100 lines or even less.

Reply:
I used the `attr` module only for validators attributes declaration. `validate`-method is only one method implemented for class-validators. It implements the basic and simple logic for scheme validation. And it can be used for many catalog structure validation at the same time. Also, it is the best way to implement nested files/catalog traversal.
So let's leave `attr` for class-instance and model initialization, data validation and conversion routines. Also, I base `RELEASE.INFO` and other content type validation mostly on the `attr` module. Because it is the good choice for that. For scheme validation (catalog/file structure validation) provided logic is simple and flexible. Inside of `validate` method I have access to all validator attributes and can use it to build complex validations. The base logic takes around 50 lines of code, other lines are docstrings and `attr` annotations.